core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java

```diff
@@ -218,7 +218,9 @@ public DiscoveryNode(String nodeName, String nodeId, String hostName, String hos
                 assert attributes.containsKey(role.getRoleName()) == false : "role name [" + role.getRoleName() + "] found in attributes";
             }
         }
-        this.roles = roles;
+        Set<Role> rolesSet = Collections.unmodifiableSet(roles);
```

Comment:
I think this `Collections.unmodifiableSet(roles);` is unnecessary and we can just do `this.roles.addAll(roles)` ?